crates/worker/src/partition/state_machine/mod.rs

```diff
@@ -430,148 +423,177 @@ impl<Codec: RawEntryCodec> StateMachine<Codec> {
-                        ctx.is_leader,
-                        restate.idempotency.id = ?idempotency_id,
-                        "Invocation is a duplicate"
-                    );
+                // We might still need to deduplicate based on the idempotency table for old invocation ids
+                if has_idempotency_key {
+                    let idempotency_id = service_invocation
+                        .compute_idempotency_id()
+                        .expect("Idempotency key must be present");
+
+                    if let Some(idempotency_metadata) = ctx.storage.get_idempotency_metadata(&idempotency_id).await? {
+                        is = ctx.get_invocation_status(&idempotency_metadata.invocation_id).await?;
+                    }
+                }
+                // Or on lock status for workflow runs with old invocation ids
+                if is_workflow_run {
+                    let keyed_service_id = service_invocation
+                        .invocation_target
+                        .as_keyed_service_id()
+                        .expect("When the handler type is Workflow, the invocation target must have a key");
 
-                    // Invocation was either resolved, or the sink was enqueued. Nothing else to do here.
-                    return Ok(None);
+                    if let VirtualObjectStatus::Locked(locked_invocation_id) = ctx
+                        .storage
+                        .get_virtual_object_status(&keyed_service_id)
+                        .await? {
+                        is = ctx.get_invocation_status(&locked_invocation_id).await?;
+                    }
                 }
+                Ok(is)
+            }
 
-                debug_if_leader!(
-                        ctx.is_leader,
-                        restate.idempotency.id = ?idempotency_id,
-                        "First time we see this idempotency id, invocation will be processed");
+        }.await?;
+
+        if previous_invocation_status == InvocationStatus::Free {
+            // --- New invocation
+            debug_if_leader!(
+                ctx.is_leader,
+                "First time we see this invocation id, invocation will be processed"
+            );
 
-                // Idempotent invocation needs to be processed for the first time, let's roll!
+            // Store the invocation id mapping if we have to and continue the processing
+            if is_workflow_run && !self.disable_idempotency_table {
+                ctx.storage
+                        .put_virtual_object_status(
+                            &service_invocation
+                                .invocation_target
+                                .as_keyed_service_id()
+                                .expect("When the handler type is Workflow, the invocation target must have a key"),
+                            &VirtualObjectStatus::Locked(invocation_id),
+                        )
+                        .await;
```

Comment:
Why do we no longer need to store the virtual object status if we are a workflow and have the idempotency table disabled?